### PR TITLE
[BEAM-1166] Update docs for getDefaultOutputCoder

### PR DESCRIPTION
**Pull request by ssisk**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [X] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [X] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [X] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

I may not have the docs entirely correct - let me know if there's a better way to phrase this.

R @jkff 
cc @dhalperi

### CODE REVIEW

**Comment by jkff** [comment]
Does the rest of the framework actually handle a `CannotProvideCoderException` thrown by this method?

**Comment by ssisk** [comment]
hrm. No, the rest of the code does not handle a CannotProvideCoderException.

I had thought based on my notes/discussion at the time when I created the JIRA issue that this was basically a documentation issue and that from a logical perspective it'd be okay to let coder inference happen later, but adding "throws CannotProviderCoderException" to source's getDefaultOutputCoder shows a couple holes. I'm guessing this is one of those changes where the changes will ripple out a bit (eg https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java#L108 and https://github.com/apache/beam/blob/master/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java#L179 - that second one makes it seem like we're just moving the "I need a coder registry" problem to a different location.)

I can definitely do the investigation/wiring/testing work if people think this is useful, relevant now and there aren't any known traps out there for me to find. It seems like for adapter sources like the HIFIO, being able to say that you don't know the coder beforehand and instead rely on coder inference would be very handy. The work to make https://github.com/apache/beam/blob/master/sdks/java/io/hadoop/input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java#L350 work  involved passing in a coder registry to the source from the outer transform, which felt dirty. It'd be nice to just check to see if there is a known coder, and throw CannotProvideCoderException if that's not the case. 

If we don't want to do this work, it may be nice to instead do the work to provide the coder registry to sources in a straightforward way.

cc @kennknowles since I know he cares about this.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Source.java [hunk 1]
```diff
@@ -63,6 +63,7 @@
 
   /**
    * Returns the default {@code Coder} to use for the data read from this source.
+   * @throws org.apache.beam.sdk.coders.CannotProvideCoderException if there is no default Coder.
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Source.java]: It is a checked exception, so I think you'd also need to change the interface here for this to be true, no?

Generally it seems correct that this should be able to throw this exception.

**Comment by ssisk** [comment]
After further review, this looks to be more complicated than originally thought. I've updated the JIRA issue with more info.

**Pull request closed without merging**